### PR TITLE
[core] Add JsonRenderer

### LINKED ISSUE #1286: [core] Export Supporting JSON Format
Hello team,

Thanks for putting together this static analyzer . While running the report i was wondering if the cmd line output supports the native **JSON** output aswell **not the one provided as CodeClimate**. 

Is this scoped for the future release ? Any directions towards the same would be much appreciated.

Thanks ,
Ashish

### PR DESCRIPTION
**Pull request by adangel**
Fixes #1286

**PR Description:**

That's a simple implementation for supporting the json format as a standard renderer.
The format is very simple, but complete.

**Open points:**
* Do we need a separate json format version? In case we change the structure of the file, a version number is needed to differenciate. Maybe the pmd version is enough for that?
* The documentation of the report formats could be expanded - I'd imagine some samples there (see https://pmd.github.io/latest/pmd_userdocs_cli_reference.html#available-report-formats)

**Sample report**
produced with the following command:
`pmd-bin-6.23.0-SNAPSHOT/bin/run.sh pmd -f json -d ~/PMD/source/pmd-github/pmd-core/src/main/java -R rulesets/java/quickstart.xml -r report.json`

[report.json.txt](https://github.com/pmd/pmd/files/4434486/report.json.txt)

Here for convienence, the beginning of the file:

```json
{
  "version": "6.23.0-SNAPSHOT",
  "timestamp": "2020-04-05T20:13:49.698+02:00",
  "files": [
    {
      "filename": "/home/andreas/PMD/source/pmd-github/pmd-core/src/main/java/net/sourceforge/pmd/PMDVersion.java",
      "violations": [
        {
          "beginline": 16,
          "begincolumn": 14,
          "endline": 79,
          "endcolumn": 1,
          "description": "The utility class name \u0027PMDVersion\u0027 doesn\u0027t match \u0027[A-Z][a-zA-Z0-9]+(Utils?|Helper)\u0027",
          "rule": "ClassNamingConventions",
          "ruleset": "Code Style",
          "priority": 1,
          "externalInfoUrl": "https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#classnamingconventions"
        }
      ]
    },
    {
      "filename": "/home/andreas/PMD/source/pmd-github/pmd-core/src/main/java/net/sourceforge/pmd/RuleContext.java",
      "violations": [
        {
          "beginline": 124,
          "begincolumn": 9,
          "endline": 125,
          "endcolumn": 111,
          "description": "Logger calls should be surrounded by log level guards.",
          "rule": "GuardLogStatement",
          "ruleset": "Best Practices",
          "priority": 2,
          "externalInfoUrl": "https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#guardlogstatement"
        }
      ]
    },
    {
      "filename": "/home/andreas/PMD/source/pmd-github/pmd-core/src/main/java/net/sourceforge/pmd/RuleSet.java",
      "violations": [
        {
          "beginline": 145,
          "begincolumn": 21,
          "endline": 146,
          "endcolumn": 90,
          "description": "Logger calls should be surrounded by log level guards.",
          "rule": "GuardLogStatement",
          "ruleset": "Best Practices",
          "priority": 2,
          "externalInfoUrl": "https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#guardlogstatement"
        },
        {
          "beginline": 474,
          "begincolumn": 21,
          "endline": 474,
          "endcolumn": 141,
          "description": "Logger calls should be surrounded by log level guards.",
          "rule": "GuardLogStatement",
          "ruleset": "Best Practices",
          "priority": 2,
          "externalInfoUrl": "https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#guardlogstatement"
        }
      ]
    },
```

Before submitting a PR, please check that:
 - [x] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [x] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/e4339ae7ab00435b8434a94121a7a904/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by kaulonline** [comment]
Thanks, Andreas. 

The JSON looks in a format that can be further consumed. I am running a few tests to validate this further. 

Appreciate your help with this and thanks again for putting this as an evolving community-supported project.

**Comment by adangel** [comment]
Note: once #2413 is merged, we need to extend the doc about this new format.

**Comment by adangel** [comment]
Hi @kaulonline , have been able to test the format? I can also provide you a pmd binary with that feature, if that makes the test on your side easier.

**Comment by kaulonline** [comment]
Hey Andreas,

Yes, I have been able to test the feature and have compiled the binary from
the source branch.

Thank you for your followup.

Really appreciate it !

Regards,
Ashish

File: pmd-core/src/main/java/net/sourceforge/pmd/renderers/JsonRenderer.java [hunk 1]
```diff
@@ -21,6 +21,12 @@
 public class JsonRenderer extends AbstractIncrementingRenderer {
     public static final String NAME = "json";
 
+    // TODO do we make this public? It would make it possible to write eg
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/renderers/JsonRenderer.java:24]: For now, we don't need it public. We don't write a reader, just a writer, so the only important thing is, that this formatVersion ends up in the serialized json file.

**Pull request merged**